File: resources/ts/schemas/AgentIssue.ts
import { z } from "zod";

export const AgentIssueSchema = z.union([
  z.object({
    ChatTemplateDoesNotCompile: z.object({
      error: z.string(),
      template_content: z.string(),
    }),
  }),
  z.object({
    HuggingFaceCannotAcquireLock: z.string(),
  }),
  z.object({
    HuggingFaceModelDoesNotExist: z.string(),
  }),
  z.object({
    ModelCannotBeLoaded: z.string(),
  }),
  z.object({
    ModelFileDoesNotExist: z.string(),
  }),
  z.object({
    SlotCannotStart: z.object({
      error: z.string(),
      slot_index: z.number(),
    }),
  }),
  z.object({
    UnableToFindChatTemplate: z.string(),
  }),
]);

export type AgentIssue = z.infer<typeof AgentIssueSchema>;

File: src/agent_issue.rs
use serde::Deserialize;
use serde::Serialize;

use crate::agent_issue_params::ChatTemplateDoesNotCompileParams;
use crate::agent_issue_params::SlotCannotStartParams;

#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(deny_unknown_fields)]
pub enum AgentIssue {
    ChatTemplateDoesNotCompile(ChatTemplateDoesNotCompileParams),
    HuggingFaceCannotAcquireLock(String),
    HuggingFaceModelDoesNotExist(String),
    ModelCannotBeLoaded(String),
    ModelFileDoesNotExist(String),
    SlotCannotStart(SlotCannotStartParams),
    UnableToFindChatTemplate(String),
}
